### PR TITLE
Create setup script for single command full dev setup

### PR DESCRIPTION
**Pull request by texhnolyze**
# Summary
Write setup script, which allows for full setup of repository and environment for development including:
- cloning of the repo
- setup of `rosdep init`
- setup of zsh configuration and inclusion of ros/colcon aliases/functions
- setup with `make install` or without sudo `make install-no-root`
- optional install of webots
- full colcon build of all packages

## Proposed changes
For easier management of ros/colcon aliases and utility functions I have created a `ros.plugin.sh` file, which can be sourced in either `.bashrc` or `.zshrc` and have also adjusted the devcontainer to use it.

Additionally, while testing permission issues with devcontainer became apparent #357. 

## Checklist

- [x] Write documentation
- [x] Test on your machine
- [ ] Test on the robot
- [ ] Create issues for future work
- [x] Triage this PR and label it

### CODE REVIEW

File: scripts/ros.plugin.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+alias rpl='ros2 param list'
+alias rpg='ros2 param get'
+
+# colcon aliases
+alias cdc='cd $COLCON_WS'
+
+alias cba='sr; cdc; colcon build --symlink-install --continue-on-error'
+alias cbs='sr; cdc; colcon build --symlink-install --packages-select'
+alias cb='sr; cdc; colcon build --symlink-install --continue-on-error --packages-up-to'
```
Comment (timonegk) [inline scripts/ros.plugin.sh:44]: Why did you add `sr` here?

File: scripts/ros.plugin.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+alias cc='cdc; colcon clean packages --packages-select'
+alias cca='cdc; colcon clean packages'
```
Comment (timonegk) [inline scripts/ros.plugin.sh:46]: Why did you change `&&` to `;`? If the `cd` fails, we should probably not run the `colcon` commands.

File: bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst [hunk 3]
```diff
@@ -3,12 +3,50 @@ Software installation with ROS2
...
 If you are not already using Ubuntu 22.04, consider installing it on your system (perhaps as a dual boot), alternately you can run it in a virtual machine (not recommended, as recently we had some issues with it; https://www.virtualbox.org/) or use the ROS2 docker (https://github.com/timonegk/rosdocked)
 
+**TLDR**: single command setup
+------------------------------
+
+**Prerequirements**
+- running Ubuntu 22.04 environment
+- existing Github account and with SSH key added to your account
+
+If you have not previously setup any of our software stack, you can use the following command to install and setup everything in one go:
```
Comment (timonegk) [inline bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst:20]:
```suggestion
If you have not previously set up any of our software stack, you can use the following command to install and set up everything in one go:
```

File: bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst [hunk 4]
```diff
@@ -17,23 +55,32 @@ If you are not already using Ubuntu 22.04, consider installing it on your system
...
+    ros-iron-plotjuggler-ros \
+    ros-iron-rmw-cyclonedds-cpp \
+    ros-iron-rqt-robot-monitor \
+    ros-iron-rqt-runtime-monitor
+
+- To aditionally get nice colored output from colcon, you can install the following pip packages:
+
+.. code-block:: bash
+  python3 -m pip install \
+    git+https://github.com/ruffsl/colcon-clean \
```
Comment (timonegk) [inline bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst:74]:
```suggestion
```
This is the same as `python3-colcon-clean`.

File: scripts/setup.sh [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+REPO_URL="git@github.com:bit-bots/bitbots_main.git"
+SHELL_CONFIG="$(cat <<EOF
+
+# >>> bit-bots initialize >>>
+
+# Ignore some deprecation warnings
+export PYTHONWARNINGS=ignore:::setuptools.command.install,ignore:::setuptools.command.easy_install,ignore:::pkg_resources
+
+# Limit ROS 2 communication to localhost (can be overridden when needed)
+export ROS_DOMAIN_ID=24
```
Comment (timonegk) [inline scripts/setup.sh:17]: Why 24?

File: scripts/setup.sh [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    done
+}
+
+setup_rosdep() {
+    if (( has_sudo )); then
+        echo "Setting up rosdep..."
+
+        if [[ ! -f "/etc/ros/rosdep/sources.list.d/20-default.list" ]]; then
+            sudo rosdep init
+        fi
```
Comment (timonegk) [inline scripts/setup.sh:61]: Shouldn't you also run `rosdep install`?

File: scripts/ros.plugin.sh (hunk 2 above)
Comment (texhnolyze) [reply scripts/ros.plugin.sh]: Just a copy paste issue. In an earlier versions we did not use `&&`, but you are right that makes more sense.

File: scripts/ros.plugin.sh [hunk 7]
```diff
@@ -0,0 +1,50 @@
+### Aliases and functions for ros2/colcon usage. Usage for either
```
Comment (jaagut) [inline scripts/ros.plugin.sh:1]:
```suggestion
### Aliases and functions for ROS 2 and colcon usage. Usage for either
```

File: scripts/ros.plugin.sh [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+shell="$(basename "$SHELL")"
+
+rid() {
+  export ROS_DOMAIN_ID="$1"
+  echo "ROS_DOMAIN_ID set to $ROS_DOMAIN_ID"
+}
+
+# Create a function to update the argcomplete so tab completion works
+# This needs to be called every time we source something ROS 2 related
+# For this previous loading of bashcompinit is required
```
Comment (jaagut) [inline scripts/ros.plugin.sh:13]: What is this sentence?

File: scripts/setup.sh [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    cd "$COLCON_WS"
+
+    set +u
+    source "/opt/ros/${ROS_DISTRO}/setup.bash"
+
+    colcon build --symlink-install --continue-on-error
+}
+
+setup_shell_aliases() {
+    if ask_question "Do you want to setup/update the bit-bots ros2/colcon shell configuration?"; then
```
Comment (jaagut) [inline scripts/setup.sh:134]:
```suggestion
    if ask_question "Do you want to setup/update the Bit-Bots ROS 2 and colcon shell configuration?"; then
```

File: scripts/setup.sh [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        fi
+    fi
+}
+
+has_sudo=0
+if ask_question "Do you have sudo rights?"; then
+    has_sudo=1
+fi
+if (( ! has_sudo )); then
+    echo "Because, you don't have sudo rights, you need ensure all necessary ros packages are preinstalled."
```
Comment (jaagut) [inline scripts/setup.sh:157]:
```suggestion
    echo "Because, you don't have sudo rights, you need ensure all necessary ROS 2 packages are installed."
```

File: bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst [hunk 11]
```diff
@@ -3,11 +3,33 @@ Software installation with ROS2
...
 
+
 **0. Use Ubuntu 22.04**
 
 As ROS works best on Ubuntu, we are using this distribution.
 Currently, ROS2 Iron runs on Ubuntu 22.04.
-If you are not already using Ubuntu 22.04, consider installing it on your system (perhaps as a dual boot), alternately you can run it in a virtual machine (not recommended, as recently we had some issues with it; https://www.virtualbox.org/) or use the ROS2 docker (https://github.com/timonegk/rosdocked)
+If you are not already using Ubuntu 22.04, consider installing it on your system (perhaps as a dual boot), alternately you can run it in a virtual machine (not recommended, as recently we had some issues with it; https://www.virtualbox.org/), a custom ROS2 docker setup (https://github.com/timonegk/rosdocked).
+
+Alternatively you can use a devcontainer :doc:`vscode-dev-container`, with a preconfigured environment and follow those instructions, as these docs do not apply to the devcontainer.
```
[the remark this answers is not in the document]
Comment (texhnolyze) [reply bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst:13]: I mean, I had a working setup for VM before on my old laptop, but I agree.
My idea was, that support for devcontainers is integrated into vscode and I'm against forcing usage of specific editors/IDEs. But I agree that it's the most sensible, so I'll put `rosdocked` into a Notes section.

File: scripts/make_basler.sh [hunk 12]
```diff
@@ -44,7 +44,7 @@ if apt list pylon --installed | grep -q $PYLON_VERSION; then
 else
     echo "Pylon driver $PYLON_VERSION is not installed. Installing..."
     # Check if we have an internet connection
-    check_internet_connection $1
+    check_internet_connection "$1"
```
[the remark this answers is not in the document]
Comment (texhnolyze) [reply scripts/make_basler.sh:47]: I have `shellcheck` linting integrated by default.
While not needed in this specific case, the recommendation is always to double quote variables, unless there is a specific reason not to.
E.g. special strings like `*` could otherwise be interpreted as globs etc.
See: https://medium.com/@stefan.paladuta17/bash-why-should-you-put-shell-variables-always-between-quotation-marks-real-world-example-ac794dd53a84

File: scripts/ros.plugin.sh (hunk 8 above)
Comment (texhnolyze) [reply scripts/ros.plugin.sh:13]: It's 3 sentences. I've added dots ^^

File: bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst [hunk 13]
```diff
@@ -3,11 +3,33 @@ Software installation with ROS2
...
+
+**TLDR**: single command setup
```
[the remark this answers is not in the document]
Comment (texhnolyze) [reply bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst:6]: Done.

File: bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst [hunk 14]
```diff
@@ -17,23 +39,32 @@ If you are not already using Ubuntu 22.04, consider installing it on your system
...
 
 **2. Install Webots**
 
-- Navigate to https://github.com/cyberbotics/webots/releases and download the ``.deb`` file of **Webots2022b**.
-- Install it using the command ``sudo apt install ~/Downloads/webots_2022b_amd64.deb`` or similar, depending on your system setup.
+Webots is a robot simulator, which we use to simulate our robots and test our software.
+It is not strictly necessary to install it, but it is very useful for development and testing.
+If you want to install it, you can do so by running ``make webots`` in the bitbots_main repository.
 
 **3. Download our software**
```
[the remark this answers is not in the document]
Comment (texhnolyze) [reply bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst:70]: I've remove the sentence and added an `<a>` tag if that's what you meant.
I guess direct rst section linking is not possible, as it is the private docs.

File: bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst [hunk 15]
```diff
@@ -73,44 +104,34 @@ In case you are not using the bash shell, replace ``~/.bashrc`` and ``bash`` wit
...
+  if [[ -f \$COLCON_WS/src/bitbots_main/scripts/ros.plugin.sh ]]; then
+    source \$COLCON_WS/src/bitbots_main/scripts/ros.plugin.sh
+  fi
 
-  alias cba='cdc && colcon build --symlink-install --continue-on-error'
-  alias cb='cdc && colcon build --symlink-install --continue-on-error --packages-up-to'
-  alias cbs='cdc && colcon build --symlink-install --packages-select'
-  alias cc='cdc && colcon clean packages --packages-select'
-  alias cca='cdc && colcon clean packages'
+  # <<< bit-bots initialize <<<
```
[the remark this answers is not in the document]
Comment (texhnolyze) [reply bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst:133]: Actually rather
```
# >>> bit-bots initialize >>>
```
was missing to ensure, that the setup script will still work idempotently.

File: bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (Flova) [reply bitbots_misc/bitbots_docs/docs/manual/tutorials/install_software_ros2.rst]: If you don't want to use VS Code: https://github.com/devcontainers/cli

**Review by Flova — APPROVED** [review]
Looks nice :+1:
